File: core/domain/bsp/firmware.go
// Package bsp provides board support package management for ELMOS.
// It handles firmware blob discovery, downloading, caching, and validation.
package bsp

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// FirmwareMgr manages firmware blob downloads and caching.
// Blobs are stored in ~/.elmos/bsp-cache/firmware/<soc>/<blob-name>
type FirmwareMgr struct {
	cacheDir   string
	httpClient *http.Client
}

// NewFirmwareMgr creates a new FirmwareMgr with the given cache directory.
func NewFirmwareMgr(cacheDir string) *FirmwareMgr {
	return &FirmwareMgr{
		cacheDir: filepath.Join(cacheDir, "firmware"),
		httpClient: &http.Client{
			Timeout: 5 * time.Minute,
		},
	}
}

// GetBlobPath returns the local path for a firmware blob.
// Creates SoC-specific subdirectory under cacheDir.
func (fm *FirmwareMgr) GetBlobPath(soc, blobName string) string {
	return filepath.Join(fm.cacheDir, soc, blobName)
}

// IsCached returns true if the blob is already cached locally.
func (fm *FirmwareMgr) IsCached(soc, blobName string) bool {
	path := fm.GetBlobPath(soc, blobName)
	_, err := os.Stat(path)
	return err == nil
}

// EnsureBlob ensures a firmware blob is available locally.
// Downloads from URL if not cached, validates checksum if provided.
func (fm *FirmwareMgr) EnsureBlob(soc, blobName, sourceURL, expectedChecksum string) (string, error) {
	destPath := fm.GetBlobPath(soc, blobName)

	// Return from cache if exists and checksum is valid
	if fm.IsCached(soc, blobName) {
		if expectedChecksum == "" {
			return destPath, nil
		}
		if err := fm.ValidateChecksum(destPath, expectedChecksum); err == nil {
			return destPath, nil
		}
		// Checksum mismatch - re-download
	}

	// Create SoC directory
	socDir := filepath.Join(fm.cacheDir, soc)
	if err := os.MkdirAll(socDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create firmware cache dir for %s: %w", soc, err)
	}

	// Download blob
	if err := fm.downloadBlob(sourceURL, destPath); err != nil {
		return "", fmt.Errorf("failed to download firmware blob %s: %w", blobName, err)
	}

	// Validate checksum after download
	if expectedChecksum != "" {
		if err := fm.ValidateChecksum(destPath, expectedChecksum); err != nil {
			_ = os.Remove(destPath)
			return "", fmt.Errorf("firmware blob %s checksum mismatch: %w", blobName, err)
		}
	}

	return destPath, nil
}

// DownloadBlobs downloads all blobs for a given machine concurrently.
// Returns a map of blob name → local path for all successfully downloaded blobs.
func (fm *FirmwareMgr) DownloadBlobs(blobs []BlobSpec) (map[string]string, error) {
	type result struct {
		blobName string
		path     string
		err      error
	}

	results := make(chan result, len(blobs))

	// Limit concurrent downloads
	sem := make(chan struct{}, 3)

	var wg sync.WaitGroup
	for _, blob := range blobs {
		wg.Add(1)
		blob := blob // capture
		go func() {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			path, err := fm.EnsureBlob(blob.SoC, blob.Name, blob.URL, blob.SHA256)
			results <- result{blobName: blob.Name, path: path, err: err}
		}()
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	paths := make(map[string]string)
	var errs []error
	for r := range results {
		if r.err != nil {
			errs = append(errs, r.err)
		} else {
			paths[r.blobName] = r.path
		}
	}

	if len(errs) > 0 {
		return paths, fmt.Errorf("failed to download %d blob(s): %v", len(errs), errs[0])
	}

	return paths, nil
}

// ValidateChecksum computes SHA256 of a file and compares to expected.
// Expected should be a lowercase hex string.
func (fm *FirmwareMgr) ValidateChecksum(filePath, expectedSHA256 string) error {
	f, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("failed to open file for checksum: %w", err)
	}
	defer func() { _ = f.Close() }()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return fmt.Errorf("failed to read file for checksum: %w", err)
	}

	actual := hex.EncodeToString(h.Sum(nil))
	if actual != expectedSHA256 {
		return fmt.Errorf("checksum mismatch: expected %s, got %s", expectedSHA256, actual)
	}
	return nil
}

// ComputeChecksum returns the SHA256 hex string of a file.
func (fm *FirmwareMgr) ComputeChecksum(filePath string) (string, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %w", err)
	}
	defer func() { _ = f.Close() }()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", fmt.Errorf("failed to read file: %w", err)
	}

	return hex.EncodeToString(h.Sum(nil)), nil
}

// CleanSocCache removes all cached blobs for a given SoC.
func (fm *FirmwareMgr) CleanSocCache(soc string) error {
	socDir := filepath.Join(fm.cacheDir, soc)
	if err := os.RemoveAll(socDir); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("failed to clean SoC cache for %s: %w", soc, err)
	}
	return nil
}

// ListCachedBlobs returns all cached blob names for a given SoC.
func (fm *FirmwareMgr) ListCachedBlobs(soc string) ([]string, error) {
	socDir := filepath.Join(fm.cacheDir, soc)
	entries, err := os.ReadDir(socDir)
	if os.IsNotExist(err) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to list cached blobs for %s: %w", soc, err)
	}

	var blobs []string
	for _, entry := range entries {
		if !entry.IsDir() {
			blobs = append(blobs, entry.Name())
		}
	}
	return blobs, nil
}

// downloadBlob downloads a file from URL to destPath.
func (fm *FirmwareMgr) downloadBlob(url, destPath string) error {
	resp, err := fm.httpClient.Get(url)
	if err != nil {
		return fmt.Errorf("HTTP GET failed for %s: %w", url, err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d for %s", resp.StatusCode, url)
	}

	// Write to a temporary file first (atomic write)
	tmpPath := destPath + ".tmp"
	f, err := os.Create(tmpPath)
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}

	_, err = io.Copy(f, resp.Body)
	_ = f.Close()
	if err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("failed to write blob data: %w", err)
	}

	// Rename to final destination (atomic on most systems)
	if err := os.Rename(tmpPath, destPath); err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("failed to finalize blob file: %w", err)
	}

	return nil
}

// --- Known SoC Firmware Blob Registry ---

// BlobSpec describes a downloadable firmware blob.
type BlobSpec struct {
	SoC    string // e.g. "rk3588", "rk3566", "h616"
	Name   string // e.g. "rk3588_ddr.bin"
	URL    string // download URL
	SHA256 string // expected checksum (empty = skip validation)
}

// KnownFirmwareBlobs is the curated registry of SoC firmware blobs.
// URLs point to the rockchip-linux/rkbin GitHub repository.
var KnownFirmwareBlobs = map[string][]BlobSpec{
	"rk3588": {
		{
			SoC:    "rk3588",
			Name:   "rk3588_ddr_lp4_2112MHz.bin",
			URL:    "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/rk3588_ddr_lp4_2112MHz_v1.16.bin",
			SHA256: "", // Populated by registry or left for runtime validation
		},
		{
			SoC:    "rk3588",
			Name:   "rk3588_spl_loader.bin",
			URL:    "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/rk3588_spl_loader_v1.15.113.bin",
			SHA256: "",
		},
	},
	"rk3566": {
		{
			SoC:  "rk3566",
			Name: "rk3566_ddr_1056MHz.bin",
			URL:  "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/rk3566_ddr_1056MHz_v1.21.bin",
		},
		{
			SoC:  "rk3566",
			Name: "rk356x_spl_loader.bin",
			URL:  "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/rk356x_spl_loader_v1.18.112.bin",
		},
	},
	"rk3399": {
		{
			SoC:  "rk3399",
			Name: "rk3399_ddr_933MHz.bin",
			URL:  "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk33/rk3399_ddr_933MHz_v1.30.bin",
		},
		{
			SoC:  "rk3399",
			Name: "rk3399_miniloader.bin",
			URL:  "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk33/rk3399_miniloader_v1.30.bin",
		},
	},
}

// GetBlobsForSoC returns all known firmware blobs for a given SoC identifier.
// Returns nil if the SoC is unknown.
func GetBlobsForSoC(soc string) []BlobSpec {
	return KnownFirmwareBlobs[soc]
}

// GetBlobByName looks up a specific blob within a SoC's blob list.
func GetBlobByName(soc, name string) *BlobSpec {
	for _, blob := range KnownFirmwareBlobs[soc] {
		if blob.Name == name {
			b := blob
			return &b
		}
	}
	return nil
}

File: src/domain/bsp/firmware.cpp
// ============================================================================
// domain/bsp/firmware.cpp — Firmware blob management
// ============================================================================

#include "firmware.hpp"

#include <openssl/evp.h>

#include <cstdio>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <memory>
#include <sstream>

namespace elmos::domain::bsp {

namespace fs = std::filesystem;

// Known firmware blob registry
static const std::unordered_map<std::string, std::vector<BlobSpec>>& known_blobs() {
    static const std::unordered_map<std::string, std::vector<BlobSpec>> blobs = {
        {"rk3588",
         {
             {.soc = "rk3588",
              .name = "rk3588_ddr_lp4_2112MHz.bin",
              .url = "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/"
                     "rk3588_ddr_lp4_2112MHz_v1.16.bin"},
             {.soc = "rk3588",
              .name = "rk3588_spl_loader.bin",
              .url = "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/"
                     "rk3588_spl_loader_v1.15.113.bin"},
         }},
        {"rk3566",
         {
             {.soc = "rk3566",
              .name = "rk3566_ddr_1056MHz.bin",
              .url = "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/"
                     "rk3566_ddr_1056MHz_v1.21.bin"},
             {.soc = "rk3566",
              .name = "rk356x_spl_loader.bin",
              .url = "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk35/"
                     "rk356x_spl_loader_v1.18.112.bin"},
         }},
        {"rk3399",
         {
             {.soc = "rk3399",
              .name = "rk3399_ddr_933MHz.bin",
              .url = "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk33/"
                     "rk3399_ddr_933MHz_v1.30.bin"},
             {.soc = "rk3399",
              .name = "rk3399_miniloader.bin",
              .url = "https://github.com/rockchip-linux/rkbin/raw/master/bin/rk33/"
                     "rk3399_miniloader_v1.30.bin"},
         }},
    };
    return blobs;
}

auto get_blobs_for_soc(const std::string& soc) -> std::vector<BlobSpec> {
    auto& blobs = known_blobs();
    auto it = blobs.find(soc);
    if (it == blobs.end())
        return {};
    return it->second;
}

auto get_blob_by_name(const std::string& soc, const std::string& name) -> const BlobSpec* {
    auto& blobs = known_blobs();
    auto it = blobs.find(soc);
    if (it == blobs.end())
        return nullptr;
    for (const auto& b : it->second) {
        if (b.name == name)
            return &b;
    }
    return nullptr;
}

FirmwareManager::FirmwareManager(const std::string& cache_dir)
    : cache_dir_((fs::path(cache_dir) / "firmware").string()) {}

auto FirmwareManager::get_blob_path(const std::string& soc, const std::string& blob_name)
    -> std::string {
    return (fs::path(cache_dir_) / soc / blob_name).string();
}

auto FirmwareManager::is_cached(const std::string& soc, const std::string& blob_name) -> bool {
    return fs::exists(get_blob_path(soc, blob_name));
}

auto FirmwareManager::ensure_blob(const std::string& soc, const std::string& blob_name,
                                  const std::string& url, const std::string& checksum)
    -> Result<std::string> {
    auto dest = get_blob_path(soc, blob_name);

    if (is_cached(soc, blob_name)) {
        if (checksum.empty())
            return dest;
        auto r = validate_checksum(dest, checksum);
        if (r)
            return dest;
    }

    std::error_code ec;
    fs::create_directories(fs::path(dest).parent_path(), ec);
    if (ec)
        return make_error(Error::generic("cannot create firmware cache: " + ec.message()));

    auto r = download_blob(url, dest);
    if (!r)
        return make_error(r.error());

    if (!checksum.empty()) {
        auto v = validate_checksum(dest, checksum);
        if (!v) {
            fs::remove(dest);
            return make_error(v.error());
        }
    }
    return dest;
}

auto FirmwareManager::validate_checksum(const std::string& path, const std::string& expected)
    -> VoidResult {
    auto actual = compute_checksum(path);
    if (!actual)
        return make_error(actual.error());
    if (*actual != expected) {
        return make_error(
            Error::generic("checksum mismatch: expected " + expected + ", got " + *actual));
    }
    return {};
}

auto FirmwareManager::compute_checksum(const std::string& path) -> Result<std::string> {
    std::ifstream file(path, std::ios::binary);
    if (!file)
        return make_error(Error::generic("cannot open file: " + path));

    auto ctx = std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)>(EVP_MD_CTX_new(),
                                                                       EVP_MD_CTX_free);
    if (!ctx)
        return make_error(Error::generic("failed to create EVP_MD_CTX"));
    if (EVP_DigestInit_ex(ctx.get(), EVP_sha256(), nullptr) != 1)
        return make_error(Error::generic("failed to init SHA256"));

    char buf[8192];
    while (file.read(buf, sizeof(buf)) || file.gcount() > 0) {
        if (EVP_DigestUpdate(ctx.get(), buf, static_cast<size_t>(file.gcount())) != 1)
            return make_error(Error::generic("failed to update SHA256"));
    }

    unsigned char hash[EVP_MAX_MD_SIZE];
    unsigned int hash_len = 0;
    if (EVP_DigestFinal_ex(ctx.get(), hash, &hash_len) != 1)
        return make_error(Error::generic("failed to finalize SHA256"));

    std::ostringstream ss;
    for (unsigned int i = 0; i < hash_len; ++i) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(hash[i]);
    }
    return ss.str();
}

auto FirmwareManager::clean_soc_cache(const std::string& soc) -> VoidResult {
    std::error_code ec;
    fs::remove_all(fs::path(cache_dir_) / soc, ec);
    return {};
}

auto FirmwareManager::list_cached_blobs(const std::string& soc)
    -> Result<std::vector<std::string>> {
    auto dir = (fs::path(cache_dir_) / soc).string();
    std::error_code ec;
    if (!fs::exists(dir, ec))
        return std::vector<std::string>{};

    std::vector<std::string> blobs;
    for (const auto& entry : fs::directory_iterator(dir, ec)) {
        if (!entry.is_directory()) {
            blobs.push_back(entry.path().filename().string());
        }
    }
    return blobs;
}

auto FirmwareManager::download_blob(const std::string& /*url*/, const std::string& /*dest*/)
    -> VoidResult {
    // HTTP download using cpp-httplib would go here.
    // For now, return an error indicating manual download is needed.
    return make_error(
        Error::generic("network downloads not yet implemented — place blobs manually"));
}

}  // namespace elmos::domain::bsp
